### PR TITLE
Branch fix search command bug

### LINKED ISSUE #91: Fix Search Bugs
1. Invalid prefixes are not detected.

Command to replicate:
```
/search ; invalidPrefix : invalidPrefix
```

Current Behaviour:
Error Message "Failed to find Pooch Contact - Search requires a name / phone / address / email / product / employment field. 🐾" is thrown.

Expected Behaviour:
Error Message "Failed to find Pooch Contact - Pooch doesn't recognise the field 'invalidPrefix' 🐾

2. Searching for fields that are not part of the person class searched for returns the entire list.

Commands to replicate:
```
/pooch-add ; name : Pooch ; phone : 98883888 ; address : Pooch Street 32 ; email : impooch@gmail.com
/search ; employment : someRandomValue
```

Current Behaviour:
Pooch Contact 'Pooch' returned as a result.

Expected Behaviour:
No results returned.

### PR DESCRIPTION
**Pull request by Joshy837**
Fixes #91 

Made changes to command and error messages for search to align with specifications.

Fixed the issue of having inconsistent searches for specific fields with specialised persons (i.e. Staff, Supplier, Maintainer)

### CODE REVIEW

File: src/main/java/seedu/address/model/person/KeywordPredicate.java [hunk 1]
```diff
@@ -29,40 +29,111 @@ public KeywordPredicate(ArgumentMultimap keywords) {
...
     public boolean test(Person person) {
-        boolean predicate = true;
 
-        predicate &= contains(person.getName().toString(), PREFIX_NAME);
-        predicate &= contains(person.getPhone().toString(), PREFIX_PHONE);
-        predicate &= contains(person.getEmail().toString(), PREFIX_EMAIL);
-        predicate &= contains(person.getAddress().toString(), PREFIX_ADDRESS);
+        // Searches through staff only
+        if (hasField(PREFIX_SALARY) || hasField(PREFIX_EMPLOYMENT)) {
+            if (!(person instanceof Staff)) {
```
Comment (jamessinmaojun) [inline src/main/java/seedu/address/model/person/KeywordPredicate.java:35]: can shorten to `return person instanceof Staff;`

File: src/main/java/seedu/address/model/person/KeywordPredicate.java [hunk 2]
```diff
@@ -29,40 +29,111 @@ public KeywordPredicate(ArgumentMultimap keywords) {
...
-            predicate &= contains(staff.getEmployment().toString(), PREFIX_EMPLOYMENT);
+            return checkNamePhoneEmailAndAddress(staff)
+                    && checkSalary(staff)
+                    && checkEmployment(staff);
         }
 
-        if (person instanceof Supplier) {
+        // Searches through supplier only
+        if (hasField(PREFIX_PRICE) || hasField(PREFIX_PRODUCT)) {
+            if (!(person instanceof Supplier)) {
```
Comment (jamessinmaojun) [inline src/main/java/seedu/address/model/person/KeywordPredicate.java:47]: same here

File: src/main/java/seedu/address/model/person/KeywordPredicate.java [hunk 3]
```diff
@@ -29,40 +29,111 @@ public KeywordPredicate(ArgumentMultimap keywords) {
...
-            predicate &= contains(supplier.getProduct().toString(), PREFIX_PRODUCT);
+            return checkNamePhoneEmailAndAddress(supplier)
+                    && checkPrice(supplier)
+                    && checkProduct(supplier);
         }
 
-        if (person instanceof Maintainer) {
+        // Searches through maintainer only
+        if (hasField(PREFIX_SKILL) || hasField(PREFIX_COMMISSION)) {
+            if (!(person instanceof Maintainer)) {
```
Comment (jamessinmaojun) [inline src/main/java/seedu/address/model/person/KeywordPredicate.java:59]: same here

File: src/main/java/seedu/address/model/person/KeywordPredicate.java [hunk 4]
```diff
@@ -29,40 +29,111 @@ public KeywordPredicate(ArgumentMultimap keywords) {
...
 
-    boolean contains(String identifier, Prefix keyword) {
-        if (keywords.getValue(keyword).isPresent()) {
-            return identifier.toLowerCase().contains(keywords.getValue(keyword).get().toLowerCase());
+    String getValue(Prefix field) {
+        return keywords.getValue(field).get();
+    }
+
+    boolean contains(String identifier, Prefix field) {
+        if (!hasField(field)) {
```
Comment (jamessinmaojun) [inline src/main/java/seedu/address/model/person/KeywordPredicate.java:82]: same here

**Review by jamessinmaojun — APPROVED** [review]
some code quality comments, not very urgent but would be nice to have. Otherwise lgtm!

**Review by chiageng — APPROVED** [review]
LGTM

**Pull request merged**